src/liballoc/abort_adapter.rs

```diff
@@ -0,0 +1,112 @@
+//! An allocator adapter that blows up by calling `handle_alloc_error` on all errors.
+//!
+//! On one hand, concrete allocator implementations should always be written
+//! without panicking on user error and OOM to give users maximum
+//! flexibility. On the other hand, code that depends on allocation succeeding
+//! should depend on `Alloc<Err=!>` to avoid repetitively handling errors from
+//! which it cannot recover.
+//!
+//! This adapter bridges the gap, effectively allowing `Alloc<Err=!>` to be
+//! implemented by any allocator.
+
+#![unstable(feature = "allocator_api",
+            reason = "the precise API and guarantees it provides may be tweaked \
+                      slightly, especially to possibly take into account the \
+                      types being stored to make room for a future \
+                      tracing garbage collector",
+            issue = "32838")]
+
+use core::usize;
+use core::ptr::NonNull;
+
+use crate::alloc::*;
+
+/// An allocator adapter that blows up by calling `handle_alloc_error` on all errors.
+///
+/// See the [module-level documentation](../../std/abort_adapter/index.html) for more.
+#[derive(Copy, Clone, Debug, Default)]
+pub struct AbortAdapter<Alloc>(pub Alloc);
+
+impl<A: AllocHelper> AllocHelper for AbortAdapter<A> {
+    type Err = !;
+}
+
+unsafe impl<A: Alloc> Alloc for AbortAdapter<A> {
+    unsafe fn alloc(&mut self, layout: Layout) -> Result<NonNull<u8>, Self::Err> {
+        self.0.alloc(layout).or_else(|_| handle_alloc_error(layout))
```

Comment:
I feel like we should handle this in a way that includes the actual error that occured. Maybe it'd be best to bound the `Err` associated type on `Debug` or `Display`?

Reply:
Sure, but I figure we can hash out what `handle_alloc_error` should take later?

Reply:
`handle_alloc_error` is stable, so it can't be changed. But yeah, this is less important for the moment.